### PR TITLE
[FIXED JENKINS-35725] 404 page when pipeline not found

### PR DESCRIPTION
**Pull request by kzantow**
Related to issue #JENKINS-35725. 

Summary of this pull request: 
. Add `FetchStatus` to pass to components to indicate a fetch is in progress or failed
. Display stylized 404 page when an invalid pipeline URL is accessed

image

@reviewbybees

### CODE REVIEW

**Comment by i386** [comment]
Screenshot looks great 👍

**Comment by michaelneale** [comment]
some gulp errors @kzantow it looks like

**Comment by kzantow** [comment]
@michaelneale yep, lint issues now fixed

**Comment by michaelneale** [comment]
@kzantow I like this, but its a bigger change to the fetcher. I would like @scherler and @cliffmeyers to weigh in (ideally @tfennelly too) as I get the impression this may be the place that other fetch logic is placed?

**Comment by michaelneale** [comment]
Any tests possible?

**Comment by michaelneale** [comment]
need to have some kind of tests for this to be merged.

**Comment by kzantow** [comment]
@michaelneale yep, will get to this tonight

**Comment by cliffmeyers** [comment]
LGTM! Nice. 🐝

**Comment by kzantow** [comment]
@cliffmeyers would appreciate a re-bee if you have a sec (added some basic test cases / slight code modification)

**Comment by kzantow** [comment]
@cliffmeyers yes, it doesn't seem to be doing what I expect :-/

File: blueocean-dashboard/src/main/js/components/PipelinePage.jsx [hunk 1]
```diff
@@ -11,8 +13,6 @@ import {
 } from '@jenkins-cd/design-language';
 import { buildOrganizationUrl, buildPipelineUrl } from '../util/UrlUtils';
 
-const { object } = PropTypes;
```
Comment (scherler) [inline blueocean-dashboard/src/main/js/components/PipelinePage.jsx]: why?

File: blueocean-dashboard/src/main/js/components/PipelinePage.jsx [hunk 2]
```diff
@@ -44,17 +52,18 @@ export default class PipelinePage extends Component {
...
-                {React.cloneElement(this.props.children, { pipeline })}
+                {this.props.children}
             </Page>
         );
     }
 }
 
 PipelinePage.propTypes = {
-    children: object,
+    children: PropTypes.any,
```
Comment (scherler) [inline blueocean-dashboard/src/main/js/components/PipelinePage.jsx]: why not add above any? IMO the code you refactor here is much cleaner then the one you deliver instead. It is obvious that object/any is PropTypes since it is use exclusively within XXX.propTypes. I drop the redundant PropTypes

File: blueocean-dashboard/src/main/js/redux/actions.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import keymirror from 'keymirror';
 
-import fetch from 'isomorphic-fetch';
+import fetch from '../util/fetch';
```
Comment (scherler) [inline blueocean-dashboard/src/main/js/redux/actions.js]: IMO the whole fetch refactor does not belong in this PR

File: blueocean-dashboard/src/main/js/util/fetch.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    }
+    onStart(id) {
+        console.log('onStart');
+        this._active.push(id);
+    }
+    onComplete(id) {
+        const idx = this._active.indexOf(id);
+        if (idx >= 0) {
+            this._active.splice(idx, 1);
+            console.log('onComplete');
```
Comment (scherler) [inline blueocean-dashboard/src/main/js/util/fetch.js]: I guess you will remove that, right?

File: blueocean-dashboard/src/main/js/components/NotFound.jsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import { Link } from 'react-router';
+
+/**
+ * Simple component to render a fullscreen 404 page
+ */
+export default () => (
+    <Fullscreen className="not-found">
+        <div className="message-box">
+            <h3>Page not found (404)</h3>
+            <div className="message">Jenkins could not find the page you were looking for. Check the URL for errors or press the back button.</div>
```
Comment (scherler) [inline blueocean-dashboard/src/main/js/components/NotFound.jsx]: Which back button? The only button is "open Dashboard"

File: blueocean-dashboard/src/main/js/redux/actions.js (hunk 3 above)
Comment (scherler) [inline blueocean-dashboard/src/main/js/redux/actions.js]: hmm, where do you expose fetch in util? You expose fetcher as default but that is not equal fetch. Did you try to see a run detail log? that should IMO fail with that code, since it will throw an exception converting txt to json

Comment (kzantow) [inline blueocean-dashboard/src/main/js/redux/actions.js]: @scherler that's a good point, it was a long running work in progress, I'll scale back the fetch refactor, it's not really necessary

File: blueocean-dashboard/src/main/js/components/PipelinePage.jsx (hunk 2 above)
Comment (kzantow) [inline blueocean-dashboard/src/main/js/components/PipelinePage.jsx]: Hm, this is definitely a code style difference of opinion. Our files are currently mixed, which is bad. Defining the vars this way into these common english terms includes `object` and `array` which are only a case differences from the Javascript `Object` and `Array` types and alone should be enough to not do it, in my opinion. On top of it, changing the prop types often requires a 2 line change, instead of a 1-line change. It will also lead to lint issues when people forget to remove unused `PropType` definitions. It's also poor code practice to have the `const` definition so far away from the usage, if that is the preferred way to do it. So, I strongly disagree that the code before is in any way 'cleaner'.

File: blueocean-dashboard/src/main/js/components/NotFound.jsx (hunk 5 above)
Comment (kzantow) [inline blueocean-dashboard/src/main/js/components/NotFound.jsx]: ... the browser back button.

File: blueocean-dashboard/src/main/js/util/fetch.js (hunk 4 above)
Comment (kzantow) [inline blueocean-dashboard/src/main/js/util/fetch.js]: yep, good catch. :'( for missing that

**Comment by kzantow** [comment]
@scherler your comments addressed, please have a re-look

File: blueocean-dashboard/src/main/js/components/PipelinePage.jsx (hunk 2 above)
Comment (cliffmeyers) [inline blueocean-dashboard/src/main/js/components/PipelinePage.jsx]: I believe @sophistifunk had made a point as well that using the PropTypes directly via destructuring will cause problems with Flow, which we may end up integrating into our plugins once we sort out any remaining kinks in the JDL.

Comment (scherler) [inline blueocean-dashboard/src/main/js/components/PipelinePage.jsx]: as you said "code style difference of opinion". As stated it is the context, but I agree to move it just before the usage makes a lot of sense. 

Most of you points are covered by eslint rules and I am against using React.PropTypes.<type> there is no benefit in it You will repeat React.PropTypes meaningless times. 

Anyway ....

File: blueocean-dashboard/src/main/js/components/NotFound.jsx (hunk 5 above)
Comment (scherler) [inline blueocean-dashboard/src/main/js/components/NotFound.jsx]: so may be put "browser"  before the "back button" so it is clear for stupid people like me.

**Comment by scherler** [comment]
besides the browser back button ;) :bee:
